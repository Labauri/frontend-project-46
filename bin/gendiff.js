#!/usr/bin/env node

import { Command } from 'commander';
import path from 'path';
import process from 'process';
import gendiff from '../src/index.js';

const program = new Command();

program
  .name('gendiff')
  .description('Compares two configuration files and shows a difference.')
  .version('1.0.0')
  .arguments('<filepath1> <filepath2>')
  .action((filepath1, filepath2) => {
    const absolutePath1 = path.resolve('__fixtures__', filepath1);
    const absolutePath2 = path.resolve('__fixtures__', filepath2);

    const diff = gendiff(absolutePath1, absolutePath2);
    console.log(diff);
  });

program.parse(process.argv);
