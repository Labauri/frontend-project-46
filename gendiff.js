#!/usr/bin/env node

const { Command } = require('commander');
const program = new Command();

program
  .name('gendiff')
  .description('Compares two configuration files and shows a difference.')
  .version('1.0.0')
  .option('-f, --format [type]', 'output format')
  .helpOption('-h, --help', 'output usage information');

program.on('--help', () => {
  console.log();
  console.log('Usage: gendiff [options] <filepath1> <filepath2>');
  console.log();
  console.log('Compares two configuration files and shows a difference.');
  console.log();
  console.log('Options:');
  console.log('  -V, --version        output the version number');
  console.log('  -f, --format [type]  output format');
  console.log('  -h, --help           output usage information');
  console.log();
});

program.parse(process.argv);
