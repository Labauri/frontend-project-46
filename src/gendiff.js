import path from 'path';
import * as fs from 'fs';
import parser from './parser.js';
import doFormatting from './formatters/index.js';
import ast from './ast.js';

const getFullPath = (filePath) => path.resolve(process.cwd(), filePath);
const extractFormat = (filePath) => path.extname(filePath).slice(1);
const getData = (filePath) => parser(fs.readFileSync(filePath, 'utf-8'), extractFormat(filePath));

const genDiff = (filePath1, filePath2, formatName = 'stylish') => {
  const fullPath1 = getFullPath(filePath1);
  const fullPath2 = getFullPath(filePath2);
  const data1 = getData(fullPath1);
  const data2 = getData(fullPath2);
  const buildAst = ast(data1, data2);
  const formatting = doFormatting(buildAst, formatName);
  return formatting;
};
export default genDiff;
