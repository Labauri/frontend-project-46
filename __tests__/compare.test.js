// import gendiff from '../src/index.js';
// import { fileURLToPath } from 'url';
// import path from 'path';

// const __filename = fileURLToPath(import.meta.url);
// const __dirname = path.dirname(__filename);

// test('compare nested JSON files', () => {
//   const filepath1 = path.join(__dirname, '../__fixtures__/file1.json');
//   const filepath2 = path.join(__dirname, '../__fixtures__/file2.json');

//   const expected = `{
//     common: {
//       + follow: false
//         setting1: Value 1
//       - setting2: 200
//       - setting3: true
//       + setting3: null
//       + setting4: blah blah
//       + setting5: {
//             key5: value5
//         }
//         setting6: {
//             doge: {
//               - wow:
//               + wow: so much
//             }
//             key: value
//           + ops: vops
//         }
//     }
//     group1: {
//       - baz: bas
//       + baz: bars
//         foo: bar
//       - nest: {
//             key: value
//         }
//       + nest: str
//     }
//   - group2: {
//         abc: 12345
//         deep: {
//             id: 45
//         }
//     }
//   + group3: {
//         deep: {
//             id: {
//                 number: 45
//             }
//         }
//         fee: 100500
//     }
// }`;

//   const result = gendiff(filepath1, filepath2);
//   expect(result).toEqual(expected);
// });

import gendiff from '../src/index.js';
import { fileURLToPath } from 'url';
import path from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

test('compare nested JSON files (default stylish format)', () => {
  const filepath1 = path.join(__dirname, '../__fixtures__/file1.json');
  const filepath2 = path.join(__dirname, '../__fixtures__/file2.json');

  const result = gendiff(filepath1, filepath2);
  expect(result).toMatchSnapshot();
});

test('compare nested JSON files in plain format', () => {
  const filepath1 = path.join(__dirname, '../__fixtures__/file1.json');
  const filepath2 = path.join(__dirname, '../__fixtures__/file2.json');

  const result = gendiff(filepath1, filepath2, 'plain');
  expect(result).toMatchSnapshot();
});
