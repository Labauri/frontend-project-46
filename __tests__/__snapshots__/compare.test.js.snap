// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compare nested JSON files (default stylish format) 1`] = `
"  - common: {
        setting1: Value 1
        setting2: 200
        setting3: true
        setting6: {
            key: value
            doge: {
                wow: 
            }
        }
    }
  - group1: {
        baz: bas
        foo: bar
        nest: {
            key: value
        }
    }
  - group2: {
        abc: 12345
        deep: {
            id: 45
        }
    }
  + timeout: 20
  + verbose: true
  + host: hexlet.io"
`;

exports[`compare nested JSON files 1`] = `
"  - common: {
        setting1: Value 1
        setting2: 200
        setting3: true
        setting6: {
            key: value
            doge: {
                wow: 
            }
        }
    }
  - group1: {
        baz: bas
        foo: bar
        nest: {
            key: value
        }
    }
  - group2: {
        abc: 12345
        deep: {
            id: 45
        }
    }
  + timeout: 20
  + verbose: true
  + host: hexlet.io"
`;
